invoiceregistry.go

```diff
@@ -52,9 +62,176 @@ func newInvoiceRegistry(cdb *channeldb.DB) *invoiceRegistry {
 		cdb:                 cdb,
 		debugInvoices:       make(map[chainhash.Hash]*channeldb.Invoice),
 		notificationClients: make(map[uint32]*invoiceSubscription),
+		newSubscriptions:    make(chan *invoiceSubscription),
+		subscriptionCancels: make(chan uint32),
+		invoiceEvents:       make(chan *invoiceEvent),
+		quit:                make(chan struct{}),
+	}
+}
+
+// Start starts the registry and all goroutines it needs to carry out its task.
+func (i *invoiceRegistry) Start() error {
+	i.wg.Add(1)
+
+	go i.invoiceEventNotifier()
+
+	return nil
+}
+
+// Stop signals the registry for a graceful shutdown.
+func (i *invoiceRegistry) Stop() {
+	close(i.quit)
+
+	i.wg.Wait()
+}
+
+// invoiceEvent represents a new event that has modified on invoice on disk.
+// Only two event types are currently supported: newly created invoices, and
+// instance where invoices are settled.
+type invoiceEvent struct {
+	isSettle bool
+
+	invoice *channeldb.Invoice
+}
+
+// invoiceEventNotifier is the dedicated goroutine responsible for accepting
+// new notification subscriptions, cancelling old subscriptions, and
+// dispatching new invoice events.
+func (i *invoiceRegistry) invoiceEventNotifier() {
+	defer i.wg.Done()
+
+	for {
+		select {
+		// A new invoice subscription has just arrived! We'll query for
+		// any backlog notifications, then add it to the set of
+		// clients.
+		case newClient := <-i.newSubscriptions:
+			// Before we add the client to our set of active
+			// clients, we'll first attempt to deliver any backlog
+			// invoice events.
+			err := i.deliverBacklogEvents(newClient)
+			if err != nil {
+				ltndLog.Errorf("unable to deliver backlog invoice "+
+					"notifications: %v", err)
+			}
+
+			ltndLog.Infof("New invoice subscription "+
+				"client: id=%v", newClient.id)
+
+			// With the backlog notifications delivered (if any),
+			// we'll add this to our active subscriptions and
+			// continue.
+			i.notificationClients[newClient.id] = newClient
+
+		// A client no longer wishes to receive invoice notifications.
+		// So we'll remove them from the set of active clients.
+		case clientID := <-i.subscriptionCancels:
+			ltndLog.Infof("Cancelling invoice subscription for "+
+				"client=%v", clientID)
+
+			delete(i.notificationClients, clientID)
+
+		// A sub-systems has just modified the invoice state, so we'll
+		// dispatch notifications to all registered clients.
+		case event := <-i.invoiceEvents:
+			for _, client := range i.notificationClients {
+				// Before we dispatch this event, we'll check
+				// to ensure that this client hasn't already
+				// received this notification in order to
+				// ensure we don't duplicate any events.
+				invoice := event.invoice
+				switch {
+				// If we've already sent this settle event to
+				// the client, then we can skip this.
+				case event.isSettle &&
+					client.settleIndex == invoice.SettleIndex:
```

Comment:
This check (and the one below) could be made stricter by using `<=`

Reply:
Well if it's `<`, then would seem that the client has actually missed a set of events?

Reply:
If it's `<`, then that means the stream has reverted to a prior update, no?

Reply:
Just seems like if we're filtering duplicate sends at index `i`, that we should filter for all indexes `<= i`

Reply:
If we add a `<` here, then we'll just skip relevant notifications all together. Consider the case where the client's latest index is `9`, and a notification for `10` arrives. `9 <= 10` so we'll skip, instead of actually delivering.

Reply:
Gotcha, you're right. I had my left and right sides mixed up. `>=` is that I meant lol

Reply:
Fixed!